lib/resty/kafka/protocol/record.lua

```diff
@@ -0,0 +1,198 @@
+
+    -- RecordBatch contains the timestamp starting value and the
+    -- maximum value of these records.
+    local first_timestamp = resp:int64() -- firstTimestamp
+    local max_timestamp = resp:int64() -- maxTimestamp
+
+    -- These fields are intended to support idempotent messages.
+    -- The features are NYI
+    local producer_id = resp:int64() -- producerId
+    local producer_epoch = resp:int16() -- producerEpoch
+    local base_sequence = resp:int32() -- baseSequence
+
+    local record_num = resp:int32() -- [records] array length
+
+    for i = 1, record_num do
+        messages[i] = {}
+
+        -- Record decoder, refer to this documents
+        -- https://kafka.apache.org/documentation/#record
+        local len = resp:varint()
+        local message_end = resp.offset + len
+
+        -- According to the protocol, only reserved.
+        local record_attributes = resp:int8()
+
+        -- Offset of this Record from RecordBatch's base value.
+        local timestamp_delta = resp:varlong()
+        local offset_delta = resp:varint()
+
+        -- The sixth bit of isControlBatch (bit 5) in attributes has a value of 1,
+        -- i.e. the current MessageSet (RecordBatch) is a ControlBatch.
+        -- !! Not process Conrtol Batch for now, they will be skipped !!
+        if band(attributes, 0x20) > 0 then
+            resp:varint() -- keyLength skipped
+            resp:int16() -- ControlBatch version skipped
+            resp:int16() -- ControlBatch type skipped
+            goto continue
+        end
+
+        messages[i] = {
+            offset = base_offset + offset_delta,
+            timestamp = first_timestamp + timestamp_delta,
+            key = resp:varint_bytes(),
+            value = resp:varint_bytes()
+        }
+
+        -- Calculates the length of the header field by the expected end position
+        -- of the message and skips the specified number of bytes.
+        -- !! Not parse message headers for now, they will be skipped !!
+        local header_len = message_end - resp.offset
+        resp.offset = resp.offset + header_len
+        ::continue::
+    end
+
+    return messages
+end
+
+
+---------
+-- Decode the message set, a different version of decoder will be selected
+-- automatically according to the MagicByte inside the message
+-- @author bzp2010 <bzp2010@apache.org>
+function _M.message_set_decode(resp, fetch_offset)
+    local ret = {}
+    local message_set_size = resp:int32()
+
+    -- Keep parsing the message until all the data in the
+    -- current response is exhausted
+    while resp:remain() > 0 do
+        -- Get 1 byte integer after 2 byte offset
+        -- [MessageSet] message magic_byte, it contains Message version
+        local message_version = resp:peek_int(16, 1)
+
+        local messages, messages_set_info, err
+        if message_version == 0 or message_version == 1 then
+            -- old MessageSet v0 or v1
+            messages, err = _message_set_v0_1_decode(resp)
```

Comment:
better passing the `ret` as an arg? to avoid creating a new table in `_message_set_v0_1_decode`.
for better performance.

Reply:
@doujiang24 I've modified it, but with all due respect, every message parsed internally by decode is output by table, it may create many tables, so this optimization may sometimes be trivial.